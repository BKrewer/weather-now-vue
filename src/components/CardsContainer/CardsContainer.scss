.cards-container {
    align-items: center;
    background-color: transparent;
    display: flex;
    flex-direction: row;
    height: calc(100vh - 70px);
    justify-content: space-between;
    max-width: 1020px;
    width: 100%;

    .weather-card {
        max-width: 300px;
        width: 100%;
        margin: 20px;
        min-height: 230px;
    }

    @media (max-width: 768px) {
        flex-direction: column;
        height: auto;
        padding: 20px 0;
    }
}
