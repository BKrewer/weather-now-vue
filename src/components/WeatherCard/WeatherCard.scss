.weather-card {
    background-color: $white;
    border-radius: 4px;
    box-shadow: 3px 3px 8px 2px $box-shadow;
    cursor: pointer;

    &--with-error {
        cursor: auto;
    }

    &__header {
        border-bottom: 1px solid $divider;
        padding: 15px;
    }

    &__title {
        color: $dark-grey;
        font-size: 20px;
        font-weight: 500;
        margin: 0;
        text-align: center;
    }

    &__temperature {
        color: $normal-temperature;
        display: block;
        font-size: 92px;
        font-weight: 500;
        padding: 15px;
        text-align: center;

        &--cold {
            color: $cold-temperature;
        }

        &--hot {
            color: $hot-temperature;
        }
    }

    &__additional-infos {
        display: flex;
        flex-direction: row;
        padding-bottom: 10px;
    }

    &__additional-info {
        text-align: center;
        color: $grey;
        width: 50%;
        font-size: 12px;

        strong {
            font-size: 16px;
            font-weight: 700;
            color: $dark-grey;
        }

        span {
            color: $grey;
            display: block;
            font-size: 14px;
            text-transform: uppercase;
            margin-bottom: 4px;
        }
    }

    &__footer {
        background-color: $card-footer-bg;
        padding: 15px;
    }

    &__updated {
        color: $grey;
        display: block;
        font-size: 10px;
        text-align: center;
    }

    .app-loading {
        height: 175px;
        align-items: center;
    }

    .weather-error {
        height: 175px;
        align-items: center;
    }
}
