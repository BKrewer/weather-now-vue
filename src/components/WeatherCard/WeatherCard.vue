<template>
  <div
    class="weather-card"
    :class="{ 'weather-card--with-error': error }"
    @mouseover="showAddInfo = true"
    @mouseleave="showAddInfo = false"
  >
    <div class="weather-card__header">
      <h2 class="weather-card__title">
        {{ cardData.city }}, {{ cardData.country }}
      </h2>
    </div>

    <weather-error v-if="error && !loading" @newRequest="emitNewRequest" />

    <app-loading v-if="loading" />

    <div v-if="updatedAt && !loading && !error" class="weather-card__content">
      <div class="weather-card__body">
        <span class="weather-card__temperature" :class="temperatureClass">
          {{ cardData.temp }}Â°
        </span>
      </div>
      <div class="weather-card__footer">
        <div v-show="showAddInfo" class="weather-card__additional-infos">
          <div class="weather-card__additional-info">
            <span>humidity</span>
            <strong>{{ cardData.humidity }}</strong
            >%
          </div>
          <div class="weather-card__additional-info">
            <span>pressure</span>
            <strong>{{ cardData.pressure }}</strong
            >hPa
          </div>
        </div>
        <span class="weather-card__updated">Updated at {{ updatedAt | formatTime }}</span>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";
import AppLoading from "../AppLoading/AppLoading.vue";
import WeatherError from "../WeatherError/WeatherError.vue";

export default {
  components: {
    AppLoading,
    WeatherError,
  },
  props: {
    cardData: {
      type: Object,
      default: () => {},
    },
  },
  data() {
    return {
      showAddInfo: false,
    };
  },
  computed: {
    ...mapState(["loading", "error", 'updatedAt']),
    temperatureClass() {
      const temp = this.cardData.temp;

      if (temp <= 5) {
        return "weather-card__temperature--cold";
      }

      if (temp > 25) {
        return "weather-card__temperature--hot";
      }

      return "";
    },
  },
  methods: {
    emitNewRequest() {
      this.$emit("newRequest");
    },
  },
};
</script>

<style lang="scss" src="./WeatherCard.scss" />