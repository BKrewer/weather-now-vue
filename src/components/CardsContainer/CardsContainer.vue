<template>
  <section class="cards-container">
    <weather-card
      v-for="weather in weatherData"
      :key="weather.city"
      :cardData="weather"
      @newRequest="requestWeather"
    />
  </section>
</template>

<script>
import WeatherCard from "@/components/WeatherCard/WeatherCard.vue";

import { mapActions, mapState } from "vuex";

const { VUE_APP_TIMER_CACHE } = process.env;

export default {
  components: {
    WeatherCard,
  },
  computed: {
    ...mapState(["weatherData", "updatedAt"]),
  },
  mounted() {
    this.requestWeather();
  },
  watch: {
    updatedAt() {
      const intervalTimer = (VUE_APP_TIMER_CACHE * 1000 * 60) + 10;
      setInterval(() => {
        this.requestWeather();
      }, intervalTimer);
    },
  },
  methods: {
    ...mapActions(["requestWeatherData"]),
    requestWeather() {
      this.requestWeatherData("3421319,3445709,184745");
    },
  },
};
</script>

<style lang="scss" src="./CardsContainer.scss" />