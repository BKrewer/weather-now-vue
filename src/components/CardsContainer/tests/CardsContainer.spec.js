import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import CardsContainer from '@/components/CardsContainer/CardsContainer.vue'
import WeatherCard from '@/components/WeatherCard/WeatherCard.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('CardsContainer.vue', () => {

    let actions
    let store
    let state

    beforeEach(() => {
        actions = {
            requestWeatherData: jest.fn(),
        }
        state = {
            weatherData: [
                { "city": "Nuuk", "country": "GL", "temp": 5, "humidity": 93, "pressure": 997 },
                { "city": "Urubici", "country": "BR", "temp": 11, "humidity": 98, "pressure": 1020 },
                { "city": "Nairobi", "country": "KE", "temp": 12, "humidity": 67, "pressure": 1023 }
            ],
            updatedAt: 1659995683744,
        }
        store = new Vuex.Store({
            actions,
            state
        })
    })

    it('renders correctly', () => {
        const wrapper = shallowMount(CardsContainer, {
            store,
            localVue
        })
        expect(wrapper.element).toMatchSnapshot();
    })

    it('renders three cards', () => {
        const wrapper = shallowMount(CardsContainer, {
            store,
            localVue
        })

        expect(wrapper.findAllComponents(WeatherCard).length).toBe(3)
    });
})