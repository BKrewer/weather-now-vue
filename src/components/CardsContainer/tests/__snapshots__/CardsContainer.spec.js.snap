// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer.vue renders correctly 1`] = `
<section
  class="cards-container"
>
  <weather-card-stub
    carddata="[object Object]"
  />
  <weather-card-stub
    carddata="[object Object]"
  />
  <weather-card-stub
    carddata="[object Object]"
  />
</section>
`;
